### ISSUE TITLE
Updating to 1.8.1 causes numerous strange errors (random & reproducible)

### ISSUE THREAD

**Issue by detoxhby** [linq2db/linq2db#708]
After updating to 1.8.1 most of our previously working code is broken. Sometimes it happens only for repetative execution, but other times it happens on first run (NRE).

Before submitting this issue I spent a full day trying out different configurations with 1.8.0 + 1.8.1: executing tests on real SQL Server to exclude any mistakes with SQL CE but the results are the same. (even gone back on code versions to check with older implementation)

**Using 1.8.0 (or older) everything is correct**, but with 1.8.1 unexplainable things happen, like:
 - null value returned where its not possible
 - different value returned than should be in the database
 - data is not saved into the database when it should have been
 - expression eval dies with NRE

The first 3 are completely mysterious and I cannot reproduce or even identify the cause of them. There are no errors, warnings...the query executes without problem, but the result is trash.

For the last one, it happens almost exactly on the same point:
Source\Linq\Query.cs:78 (in `public Expression GetIQueryable(int n, Expression expr)`)
![image](https://cloud.githubusercontent.com/assets/1679462/26782330/559800e6-49f3-11e7-933b-1376f3bb0ed6.png)
```
Exception message: An exception of type 'System.NullReferenceException' occurred in linq2db.dll but was not handled in user code
Stack trace:
lambda_method(Closure , Expression )
LinqToDB.Linq.Query.GetIQueryable(Int32 n, Expression expr) hely: c:\dev\TEMP\linq2db\Source\Linq\Query.cs, sor: 78
lambda_method(Closure , Expression , Object[] )
LinqToDB.Linq.Query`1.SetParameters(Expression expr, Object[] parameters, Int32 idx) hely: c:\dev\TEMP\linq2db\Source\Linq\Query.cs, sor: 422
LinqToDB.Linq.Query`1.SetCommand(IDataContext dataContext, Expression expr, Object[] parameters, Int32 idx, Boolean clearQueryHints) hely: c:\dev\TEMP\linq2db\Source\Linq\Query.cs, sor: 399
LinqToDB.Linq.Query`1.<RunQuery>d__11.MoveNext() hely: c:\dev\TEMP\linq2db\Source\Linq\Query.cs, sor: 379
LinqToDB.Linq.Query`1.<Map>d__69.MoveNext() hely: c:\dev\TEMP\linq2db\Source\Linq\Query.cs, sor: 1184
```

For one example, the original LINQ expression is:
```
.Call System.Linq.Queryable.Select(
    .Call System.Linq.Queryable.Where(
        .Constant<LinqToDB.Linq.Table`1[Entities.Substitution]>(LinqToDB.Linq.Table`1[Entities.Substitution]),
        '(.Lambda #Lambda1<System.Func`2[Entities.Substitution,System.Boolean]>)),
    '(.Lambda #Lambda2<System.Func`2[Entities.Substitution,Entities.User]>))

.Lambda #Lambda1<System.Func`2[Entities.Substitution,System.Boolean]>(Entities.Substitution $substitution)
{
    .Call System.Linq.Queryable.Contains(
        .Constant<DataAccess.Implementations.UserRepository+<>c__DisplayClass13_0>(DataAccess.Implementations.UserRepository+<>c__DisplayClass13_0).subQuery,
        $substitution.Id) && $substitution.StartDate <= System.DateTime.Now && $substitution.EndDate >= System.DateTime.Now
}

.Lambda #Lambda2<System.Func`2[Entities.Substitution,Entities.User]>(Entities.Substitution $substitution)
{
    $substitution.User
}
```
... where `.Constant<DataAccess.Implementations.UserRepository+<>c__DisplayClass13_0>(DataAccess.Implementations.UserRepository+<>c__DisplayClass13_0).subQuery` is:
```
.Call System.Linq.Queryable.Select(
    .Call System.Linq.Queryable.Where(
        .Constant<LinqToDB.Linq.Table`1[.Entities.SubstitutionUser]>(LinqToDB.Linq.Table`1[.Entities.SubstitutionUser]),
        '(.Lambda #Lambda1<System.Func`2[.Entities.SubstitutionUser,System.Boolean]>)),
    '(.Lambda #Lambda2<System.Func`2[.Entities.SubstitutionUser,System.Guid]>))

.Lambda #Lambda1<System.Func`2[.Entities.SubstitutionUser,System.Boolean]>(.Entities.SubstitutionUser $substitutionUsers)
{
    $substitutionUsers.UserId == (.Constant<DataAccess.Implementations.UserRepository+<>c__DisplayClass13_0>(DataAccess.Implementations.UserRepository+<>c__DisplayClass13_0).user).Id
}

.Lambda #Lambda2<System.Func`2[.Entities.SubstitutionUser,System.Guid]>(.Entities.SubstitutionUser $substitutionUsers)
{
    $substitutionUsers.SubstitutionId
}
```
(`(.Constant<DataAccess.Implementations.UserRepository+<>c__DisplayClass13_0>(DataAccess.Implementations.UserRepository+<>c__DisplayClass13_0).user).Id` is a valid Guid value)

Real code:
```c#
var user = GetById(userID);

var subQuery = from substitutionUsers in Context.SubstitutionUsers
               where substitutionUsers.UserId == user.Id
               select substitutionUsers.SubstitutionId;

var query = from substitution in Context.Substitutions
            where subQuery.Contains(substitution.Id) &&
                    substitution.StartDate <= DateTime.Now &&
                    substitution.EndDate >= DateTime.Now
            select substitution.User;

user.SubstitutedUsers = query.ToList();
```

### Steps to reproduce
I will try to corner the cases even more but there is currently nothing to stick to as the linq2db part is running okay but produces invalid results (except the NRE above).

**I would like to ask to make investigation on the diff of 1.8.0<->1.8.1**  

We are stuck on 1.8.0 and would not be able to upgrade ever in the future which is a complete disaster for our long term project. I would be glad if we'd figure out what happened and how to solve it. (will provide any assistance necessary to get over it)

_Extra info_: we are using dynamically compiled classes in some queries. Are there any change on assembly loading/identification in this patch? (to clarify: problems occour even if no dynamic assemblies involved)

### Environment details
linq2db version: *1.8.1*  
Database Server: *SQL Server 2014 SP1* and *SQL Server CE 4.0*  
Database Provider: *SqlServer* and *SqlCe*  
Operating system: *Windows 10*  
Framework version: *.NET Framework 4.6.2*

**Comment by ili** [comment]
@detoxhby thanks for feedback!

Can you please generate failing test, as described [here](https://github.com/linq2db/linq2db/wiki/Issue-reporting#generating-the-test)?

**Comment by detoxhby** [comment]
@ili [https://pastebin.com/wzJpkk4E](https://pastebin.com/wzJpkk4E)

**Comment by ili** [comment]
@detoxhby thanks! does this generated with 1.8.1?
If yes, can you please generate with 1.8.0?

**Comment by detoxhby** [comment]
@ili https://pastebin.com/D4hudgEs

Seems the same for 1.8.0 ( _except the fact query executes correctly :)_ )
Anyway, for the association `ThisKey` would be `P1` and `OtherKey` would be `T5`.`P1` but its missing from both generated file.

**Comment by ili** [comment]
@detoxhby thanks again

**Comment by sdanyliv** [comment]
@detoxhby, try to set the following setting
```cs
Configuration.Linq.UseBinaryAggregateExpression = false;
```

**Comment by detoxhby** [comment]
@sdanyliv thanks, it solved every other problems as well!

Info: with this setting in place I experience 16-18% performance degradation on the full test suite runtime. This may or may not come from other changes in 1.8.1, needs further testing.

**Comment by ili** [comment]
We have released 1.8.2 as hotfix for this problem.
